executor/write.go

```diff
@@ -886,15 +886,56 @@ func (e *InsertValues) getRows(cols []*table.Column, ignoreErr bool) (rows [][]t
 }
 
 func (e *InsertValues) getRow(cols []*table.Column, list []expression.Expression, ignoreErr bool) ([]types.Datum, error) {
-	vals := make([]types.Datum, len(list))
+	row := make([]types.Datum, len(e.Table.Cols()))
+	hasValue := make([]bool, len(e.Table.Cols()))
+	if err := e.fillDefaultValues(row, ignoreErr); err != nil {
+		return nil, errors.Trace(err)
+	}
+
 	for i, expr := range list {
-		val, err := expr.Eval(nil)
-		vals[i] = val
+		val, err := expr.Eval(row)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		val, err = table.CastValue(e.ctx, val, cols[i].ToInfo())
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+
+		offset := cols[i].Offset
+		row[offset] = val
+		hasValue[offset] = true
 		if err != nil {
 			return nil, errors.Trace(err)
 		}
 	}
-	return e.fillRowData(cols, vals, ignoreErr)
+
+	return e.checkRowData(cols, len(list), hasValue, row, ignoreErr)
+}
+
+func (e *InsertValues) fillDefaultValues(row []types.Datum, ignoreErr bool) error {
```

Comment:
hasValue columns is filled by default values too?

Reply:
`fillDefaultValues` invoked before eval actual insert expression. It fill a empty row with default value, then this row role as a eval context.